File: edge/src/protocol/frames.ts
/**
 * Wire protocol v1 frame codec (protocol/PROTOCOL.md §2).
 *
 *   frame := type:u8 | stream_id:u32 (big-endian) | payload:bytes
 *
 * `decodeFrame` enforces every connection-level "malformed" rule of §2.2 and throws
 * `ProtocolError` ("malformed" → close 1002, "stream_error" → per-stream handling).
 * Field-level schema checks of JSON payloads live with the consumers (tunnel-object).
 */

export const FrameType = {
  HELLO: 0x01,
  READY: 0x02,
  GOAWAY: 0x03,
  DRAIN: 0x04,
  REQ_HEAD: 0x10,
  REQ_BODY: 0x11,
  REQ_END: 0x12,
  RES_HEAD: 0x20,
  RES_BODY: 0x21,
  RES_END: 0x22,
  WS_TEXT: 0x30,
  WS_BINARY: 0x31,
  WS_CLOSE: 0x32,
  WINDOW: 0x40,
  RESET: 0x41,
  ACK: 0x42,
} as const;
export type FrameType = (typeof FrameType)[keyof typeof FrameType];

/** Normative wire constants (§2.2). Edge *policy* timeouts live in lib/config.ts. */
export const HEADER_SIZE = 5;
export const MAX_BODY_CHUNK = 64 * 1024;
export const MAX_WS_MESSAGE = 1024 * 1024;
export const MAX_JSON_PAYLOAD = 128 * 1024;
export const STREAM_WINDOW = 1024 * 1024;
export const CONN_WINDOW = 8 * 1024 * 1024;
export const MAX_WINDOW = 2 ** 31 - 1;
export const WS_MSG_CREDIT = 8;
export const WS_MSG_TOLERANCE = 16;
export const MAX_STREAMS = 128;
export const CREDIT_TIMEOUT_MS = 30_000;
export const MAX_STREAM_ID = 2 ** 32 - 1;

/** Heartbeat text messages (§7): the only text allowed on the agent socket. */
export const PING = "tuzy-ping";
export const PONG = "tuzy-pong";

const STREAM0_TYPES = new Set<number>([FrameType.HELLO, FrameType.READY, FrameType.GOAWAY, FrameType.DRAIN]);
const JSON_TYPES = new Set<number>([
  FrameType.HELLO,
  FrameType.READY,
  FrameType.GOAWAY,
  FrameType.REQ_HEAD,
  FrameType.RES_HEAD,
  FrameType.WS_CLOSE,
  FrameType.RESET,
]);
const EMPTY_TYPES = new Set<number>([FrameType.DRAIN, FrameType.REQ_END, FrameType.RES_END]);
const U32_TYPES = new Set<number>([FrameType.WINDOW, FrameType.ACK]);
const KNOWN_TYPES = new Set<number>(Object.values(FrameType));

export type ProtocolErrorKind = "malformed" | "stream_error";

export class ProtocolError extends Error {
  constructor(
    readonly kind: ProtocolErrorKind,
    message: string,
    /** Stream the error applies to (stream_error only). */
    readonly streamId = 0,
    /** Visitor-WS close code for stream errors: 1007 invalid UTF-8, 1009 too big. */
    readonly closeCode = 1002,
  ) {
    super(message);
    this.name = "ProtocolError";
  }
}

/** A decoded frame. Exactly one payload view is meaningful, depending on `type`. */
export interface Frame {
  type: FrameType;
  streamId: number;
  /** Raw payload bytes: a view into the (per-message, never reused) WebSocket buffer. */
  payload: Uint8Array;
  /** Parsed JSON object for JSON frame types. */
  json?: Record<string, unknown>;
  /** Credit value for WINDOW / ACK. */
  value?: number;
}

const utf8Strict = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });
const utf8 = new TextEncoder();

/** Decodes one binary WebSocket message into a frame, validating §2.2. */
export function decodeFrame(message: ArrayBuffer | ArrayBufferView): Frame {
  // ArrayBuffer.isView (not `instanceof ArrayBuffer`): buffers may come from another realm.
  const bytes = ArrayBuffer.isView(message)
    ? new Uint8Array(message.buffer, message.byteOffset, message.byteLength)
    : new Uint8Array(message);
  if (bytes.byteLength < HEADER_SIZE) throw new ProtocolError("malformed", "frame shorter than header");

  const type = bytes[0]!;
  if (!KNOWN_TYPES.has(type)) throw new ProtocolError("malformed", `unknown frame type 0x${type.toString(16)}`);
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const streamId = view.getUint32(1);
  const payload = bytes.subarray(HEADER_SIZE);

  if (STREAM0_TYPES.has(type)) {
    if (streamId !== 0) throw new ProtocolError("malformed", "control frame on non-zero stream");
  } else if (type !== FrameType.WINDOW && streamId === 0) {
    throw new ProtocolError("malformed", "stream frame on stream 0");
  }

  const frame: Frame = { type: type as FrameType, streamId, payload };

  if (EMPTY_TYPES.has(type) && payload.byteLength !== 0) {
    throw new ProtocolError("malformed", "payload must be empty");
  }
  if (U32_TYPES.has(type)) {
    if (payload.byteLength !== 4) throw new ProtocolError("malformed", "credit payload must be 4 bytes");
    const value = view.getUint32(HEADER_SIZE);
    if (value === 0) throw new ProtocolError("malformed", "credit value must be > 0");
    frame.value = value;
  }
  if (JSON_TYPES.has(type)) {
    if (payload.byteLength > MAX_JSON_PAYLOAD) throw new ProtocolError("malformed", "JSON payload too large");
    let parsed: unknown;
    try {
      parsed = JSON.parse(utf8Strict.decode(payload));
    } catch {
      throw new ProtocolError("malformed", "invalid JSON payload");
    }
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new ProtocolError("malformed", "JSON payload must be an object");
    }
    frame.json = parsed as Record<string, unknown>;
  }
  if ((type === FrameType.REQ_BODY || type === FrameType.RES_BODY) && payload.byteLength > MAX_BODY_CHUNK) {
    throw new ProtocolError("malformed", "body chunk larger than 64 KiB");
  }
  if (type === FrameType.WS_TEXT || type === FrameType.WS_BINARY) {
    if (payload.byteLength > MAX_WS_MESSAGE) throw new ProtocolError("stream_error", "ws message too big", streamId, 1009);
    if (type === FrameType.WS_TEXT) {
      try {
        utf8Strict.decode(payload);
      } catch {
        throw new ProtocolError("stream_error", "ws text is not valid UTF-8", streamId, 1007);
      }
    }
  }
  return frame;
}

/** Encodes a frame with a raw payload. */
export function encodeFrame(type: FrameType, streamId: number, payload: Uint8Array = new Uint8Array(0)): Uint8Array {
  const out = new Uint8Array(HEADER_SIZE + payload.byteLength);
  out[0] = type;
  new DataView(out.buffer).setUint32(1, streamId);
  out.set(payload, HEADER_SIZE);
  return out;
}

/** Encodes a JSON frame (compact JSON). */
export function encodeJsonFrame(type: FrameType, streamId: number, value: Record<string, unknown>): Uint8Array {
  return encodeFrame(type, streamId, utf8.encode(JSON.stringify(value)));
}

/** Encodes a WINDOW / ACK frame. */
export function encodeU32Frame(type: typeof FrameType.WINDOW | typeof FrameType.ACK, streamId: number, value: number): Uint8Array {
  const payload = new Uint8Array(4);
  new DataView(payload.buffer).setUint32(0, value);
  return encodeFrame(type, streamId, payload);
}

/** Encodes a WS_TEXT frame from a string. */
export function encodeTextFrame(streamId: number, text: string): Uint8Array {
  return encodeFrame(FrameType.WS_TEXT, streamId, utf8.encode(text));
}

/** Decodes a WS_TEXT payload (already validated by decodeFrame). */
export function payloadText(payload: Uint8Array): string {
  return utf8Strict.decode(payload);
}

File: internal/protocol/frames.go
// Package protocol implements the tuzy wire protocol v1 frame codec (protocol/PROTOCOL.md §2).
//
//	frame := type:u8 | stream_id:u32 (big-endian) | payload:bytes
//
// Decode enforces every connection-level "malformed" rule of §2.2 (the receiver closes the agent
// socket with 1002) and reports per-stream WebSocket errors as KindStreamError.
package protocol

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"unicode/utf8"
)

// FrameType is the first byte of every frame.
type FrameType uint8

// Frame types (§2.1).
const (
	Hello    FrameType = 0x01
	Ready    FrameType = 0x02
	Goaway   FrameType = 0x03
	Drain    FrameType = 0x04
	ReqHead  FrameType = 0x10
	ReqBody  FrameType = 0x11
	ReqEnd   FrameType = 0x12
	ResHead  FrameType = 0x20
	ResBody  FrameType = 0x21
	ResEnd   FrameType = 0x22
	WSText   FrameType = 0x30
	WSBinary FrameType = 0x31
	WSClose  FrameType = 0x32
	Window   FrameType = 0x40
	Reset    FrameType = 0x41
	Ack      FrameType = 0x42
)

var typeNames = map[FrameType]string{
	Hello: "HELLO", Ready: "READY", Goaway: "GOAWAY", Drain: "DRAIN",
	ReqHead: "REQ_HEAD", ReqBody: "REQ_BODY", ReqEnd: "REQ_END",
	ResHead: "RES_HEAD", ResBody: "RES_BODY", ResEnd: "RES_END",
	WSText: "WS_TEXT", WSBinary: "WS_BINARY", WSClose: "WS_CLOSE",
	Window: "WINDOW", Reset: "RESET", Ack: "ACK",
}

func (t FrameType) String() string {
	if n, ok := typeNames[t]; ok {
		return n
	}
	return fmt.Sprintf("0x%02x", uint8(t))
}

// Normative wire constants (§2.2). Edge policy timeouts are not part of the protocol.
const (
	HeaderSize     = 5
	MaxBodyChunk   = 64 * 1024
	MaxWSMessage   = 1024 * 1024
	MaxJSONPayload = 128 * 1024
	StreamWindow   = 1024 * 1024
	ConnWindow     = 8 * 1024 * 1024
	MaxWindow      = 1<<31 - 1
	WSMsgCredit    = 8
	WSMsgTolerance = 16
	MaxStreams     = 128
)

// Heartbeat text messages (§7).
const (
	Ping = "tuzy-ping"
	Pong = "tuzy-pong"
)

// ErrorKind classifies decode failures.
type ErrorKind int

const (
	// KindMalformed is a connection error: close the agent socket with 1002.
	KindMalformed ErrorKind = iota
	// KindStreamError affects one visitor WebSocket only (1007 / 1009).
	KindStreamError
)

// Error is a protocol violation found while decoding.
type Error struct {
	Kind      ErrorKind
	StreamID  uint32
	CloseCode int // for KindStreamError: 1007 (invalid UTF-8) or 1009 (too big)
	Msg       string
}

func (e *Error) Error() string { return "protocol: " + e.Msg }

func malformed(format string, a ...any) error {
	return &Error{Kind: KindMalformed, Msg: fmt.Sprintf(format, a...)}
}

// IsMalformed reports whether err is a connection-level protocol error.
func IsMalformed(err error) bool {
	var pe *Error
	return errors.As(err, &pe) && pe.Kind == KindMalformed
}

// Frame is a decoded frame. Payload aliases the input buffer.
type Frame struct {
	Type     FrameType
	StreamID uint32
	Payload  []byte
	// Value is the credit for WINDOW / ACK.
	Value uint32
}

func (t FrameType) known() bool { _, ok := typeNames[t]; return ok }

func (t FrameType) stream0Only() bool { return t == Hello || t == Ready || t == Goaway || t == Drain }

// IsJSON reports whether the payload of t is a JSON object.
func (t FrameType) IsJSON() bool {
	switch t {
	case Hello, Ready, Goaway, ReqHead, ResHead, WSClose, Reset:
		return true
	}
	return false
}

func (t FrameType) mustBeEmpty() bool { return t == Drain || t == ReqEnd || t == ResEnd }

func (t FrameType) isU32() bool { return t == Window || t == Ack }

// Decode parses one binary WebSocket message, validating §2.2.
func Decode(msg []byte) (Frame, error) {
	if len(msg) < HeaderSize {
		return Frame{}, malformed("frame shorter than header")
	}
	t := FrameType(msg[0])
	if !t.known() {
		return Frame{}, malformed("unknown frame type 0x%02x", msg[0])
	}
	f := Frame{Type: t, StreamID: binary.BigEndian.Uint32(msg[1:5]), Payload: msg[HeaderSize:]}

	if t.stream0Only() {
		if f.StreamID != 0 {
			return Frame{}, malformed("%s on non-zero stream", t)
		}
	} else if t != Window && f.StreamID == 0 {
		return Frame{}, malformed("%s on stream 0", t)
	}
	if t.mustBeEmpty() && len(f.Payload) != 0 {
		return Frame{}, malformed("%s payload must be empty", t)
	}
	if t.isU32() {
		if len(f.Payload) != 4 {
			return Frame{}, malformed("%s payload must be 4 bytes", t)
		}
		f.Value = binary.BigEndian.Uint32(f.Payload)
		if f.Value == 0 {
			return Frame{}, malformed("%s value must be > 0", t)
		}
	}
	if t.IsJSON() {
		if len(f.Payload) > MaxJSONPayload {
			return Frame{}, malformed("JSON payload too large")
		}
		if !isJSONObject(f.Payload) {
			return Frame{}, malformed("%s payload must be a JSON object", t)
		}
	}
	if (t == ReqBody || t == ResBody) && len(f.Payload) > MaxBodyChunk {
		return Frame{}, malformed("body chunk larger than 64 KiB")
	}
	if t == WSText || t == WSBinary {
		if len(f.Payload) > MaxWSMessage {
			return Frame{}, &Error{Kind: KindStreamError, StreamID: f.StreamID, CloseCode: 1009, Msg: "ws message too big"}
		}
		if t == WSText && !utf8.Valid(f.Payload) {
			return Frame{}, &Error{Kind: KindStreamError, StreamID: f.StreamID, CloseCode: 1007, Msg: "ws text is not valid UTF-8"}
		}
	}
	return f, nil
}

// isJSONObject reports whether b is exactly one valid JSON object (no trailing data).
func isJSONObject(b []byte) bool {
	if !utf8.Valid(b) {
		return false
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(b, &obj); err != nil {
		return false
	}
	return obj != nil // `null` unmarshals into a nil map
}

// Encode builds a frame with a raw payload.
func Encode(t FrameType, streamID uint32, payload []byte) []byte {
	out := make([]byte, HeaderSize+len(payload))
	out[0] = byte(t)
	binary.BigEndian.PutUint32(out[1:5], streamID)
	copy(out[HeaderSize:], payload)
	return out
}

// EncodeJSON builds a JSON frame from v (a struct or map).
func EncodeJSON(t FrameType, streamID uint32, v any) ([]byte, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	return Encode(t, streamID, b), nil
}

// EncodeU32 builds a WINDOW or ACK frame.
func EncodeU32(t FrameType, streamID, value uint32) []byte {
	var p [4]byte
	binary.BigEndian.PutUint32(p[:], value)
	return Encode(t, streamID, p[:])
}

// UnmarshalPayload decodes a JSON frame payload into v.
func (f Frame) UnmarshalPayload(v any) error {
	if err := json.Unmarshal(f.Payload, v); err != nil {
		return &Error{Kind: KindMalformed, Msg: fmt.Sprintf("bad %s payload: %v", f.Type, err)}
	}
	return nil
}
